app/org/sagebionetworks/bridge/services/ScheduledActivityService.java

```diff
@@ -181,12 +181,86 @@ public ScheduledActivityList getActivityHistory(String healthCode,
+        ScheduleContext newContext = new ScheduleContext.Builder().withContext(context).withEvents(events).build();
+        
+        // Get scheduled activities, persisted activities, and compare them
+        List<ScheduledActivity> scheduledActivities = scheduleActivitiesForPlans(newContext);
+        List<ScheduledActivity> dbActivities = activityDao.getActivities(newContext.getEndsOn().getZone(), scheduledActivities);
+        
+        Map<String, ScheduledActivity> dbMap = Maps.uniqueIndex(dbActivities, ScheduledActivity::getGuid);
+        List<ScheduledActivity> saves = Lists.newArrayList();
+        for (int i=0; i < scheduledActivities.size(); i++) {
+            ScheduledActivity activity = scheduledActivities.get(i);
+            ScheduledActivity dbActivity = dbMap.get(activity.getGuid());
+            
+            V3_MERGE.mergeActivityLists(saves, scheduledActivities, activity, dbActivity, i);
+        }        
+        activityDao.saveActivities(saves);
+        
+        return orderActivities(scheduledActivities, V3_FILTER);
     }
     
     public List<ScheduledActivity> getScheduledActivitiesV4(ScheduleContext context) {
         checkNotNull(context);
-        return getScheduledActivities(context, V4_FILTER, V4_MERGE);
+        
+        Validate.nonEntityThrowingException(VALIDATOR, context);
+        
+        // Add events for scheduling
+        Map<String, DateTime> events = createEventsMap(context);
+        ScheduleContext newContext = new ScheduleContext.Builder().withContext(context).withEvents(events).build();
+        
+        // Get scheduled activities, persisted activities, and compare them
+        List<ScheduledActivity> scheduledActivities = scheduleActivitiesForPlans(newContext);
+
+        // Get all persisted activities within the time frame, not just those found by the scheduler (as in v3).
+        Set<String> activityGuids = getScheduleActivityGuids(scheduledActivities);
+        Map<String, ScheduledActivity> dbMap = retrieveAllPersistedActivitiesIntoMap(context, activityGuids);
+        
+        List<ScheduledActivity> saves = Lists.newArrayList();
+        for (int i=0; i < scheduledActivities.size(); i++) {
+            ScheduledActivity activity = scheduledActivities.get(i);
+            ScheduledActivity dbActivity = dbMap.remove(activity.getGuid());
+            
+            V4_MERGE.mergeActivityLists(saves, scheduledActivities, activity, dbActivity, i);
+        }
+        activityDao.saveActivities(saves);
+        
+        // We've removed all the persisted activities that were found by the scheduler. Those have been saved
+        // as necessary. The remaining tasks were scheduled based on user-triggered events that can be updated, 
+        // not events like "enrollment" but events like "when this activity is finished." These are now all 
+        // added to the activities that will be return.
+        scheduledActivities.addAll(dbMap.values());
```

Comment:
V3 and V4 are almost identical, differing in only a few places. How much of a breaking change would it be to have V3 also return all previous activities for a guid, so that we don't have to have two methods that are nearly identical? (Given that we're also reverting V4_MERGE to match V3_MERGE, the only functional difference between V3 and V4 is that V4 also shows finished and expired tasks.)

If we decide that showing historical tasks is a breaking change for V3, can we add a post-processing step? This would be cleaner than having two nearly identical methods.